torch/nn/modules/adaptive.py

```diff
@@ -0,0 +1,207 @@
+            osz = self.cutoffs[i + 1] - self.cutoffs[i]
+
+            projection = Sequential(
+                Linear(self.in_features, hsz, bias=False),
+                Linear(hsz, osz)
+            )
+
+            self.tail.append(projection)
+
+    def reset_parameters(self):
+        self.head.reset_parameters()
+        for i2h, h2o in self.tail:
+            i2h.reset_parameters()
+            h2o.reset_parameters()
+
+    def forward(self, input, target):
+        if input.size(0) != target.size(0):
+            raise RuntimeError('Input and target should have the same size '
+                               'in the batch dimension.')
+
+        used_rows = 0
+        batch_size = target.size(0)
+        out_size = batch_size if (self.return_logprob) else 1
+
+        output = input.new(out_size).zero_()
+        gather_inds = target.new(batch_size).zero_()
+
+        cutoff_values = [0] + self.cutoffs
+        for i in range(len(cutoff_values) - 1):
+
+            low_idx = cutoff_values[i]
+            high_idx = cutoff_values[i + 1]
+
+            target_mask = (target >= low_idx) & (target < high_idx)
+            row_indices = target_mask.nonzero().squeeze()
+
+            if row_indices.numel() == 0:
+                continue
+
+            relative_target = target[target_mask] - cutoff_values[i]
+
+            if i == 0:
+                gather_inds.index_copy_(0, row_indices, relative_target)
+
+            else:
+                input_subset = input.index_select(0, row_indices)
+                cluster_output = self.tail[i - 1](input_subset)
+                cluster_index = self.shortlist_size + i - 1
+
+                gather_inds.index_fill_(0, row_indices, cluster_index)
+
+                if self.return_logprob:
+                    cluster_logprob = log_softmax(cluster_output, dim=1)
+                    local_logprob = cluster_logprob.gather(1, relative_target.unsqueeze(1))
+                    output.index_copy_(0, row_indices, local_logprob.squeeze(1))
+
+                else:
+                    output += cross_entropy(cluster_output, relative_target, size_average=False)
+
+            used_rows += row_indices.numel()
+
+        if used_rows != batch_size:
+            raise RuntimeError("Target values should be in [0, {}], "
+                               "but values in range [{}, {}] "
+                               "were found. ".format(self.n_classes - 1,
+                                                     int(target.min()),
+                                                     int(target.max())))
+
+        head_output = self.head(input)
+
+        if self.return_logprob:
+            head_logprob = log_softmax(head_output, dim=1)
+            output += head_logprob.gather(1, gather_inds.unsqueeze(1)).squeeze()
+
+        else:
+            output += cross_entropy(head_output, gather_inds, size_average=False)
+            output /= batch_size
+
+        return output
+
+    def get_log_proba(self, input):
```

Comment:
This should read `log_prob` or `log_probs` to be consistent with other places in PyTorch that use log probability (e.g. distributions).